// @ui-schema/ds-material/esm/Component/ListButton/ListButton.js
import { AccessTooltipIcon } from '@ui-schema/ds-material';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';
import React from 'react';
export const ListButton = ({
  onClick,
  Icon,
  title,
  btnSize,
  btnVariant,
  btnColor,
  style,
  showLabel
}) => {
  return /*#__PURE__*/React.createElement(React.Fragment, null, !showLabel && Icon ? /*#__PURE__*/React.createElement(IconButton, {
    onClick: onClick,
    size: btnSize,
    color: btnColor,
    style: style
  }, /*#__PURE__*/React.createElement(AccessTooltipIcon, {
    title: title
  }, /*#__PURE__*/React.createElement(Icon, {
    fontSize: 'inherit',
    style: {
      margin: btnSize === 'small' ? 2 : undefined
    }
  }))) : /*#__PURE__*/React.createElement(Button, {
    onClick: onClick,
    size: btnSize,
    color: btnColor,
    variant: btnVariant,
    style: style,
    startIcon: Icon ? /*#__PURE__*/React.createElement(Icon, {
      fontSize: 'inherit'
    }) : undefined
  }, title));
};
