// @ui-schema/ui-schema/UIGenerator/UIGenerator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UIProvider = exports.UIGenerator = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _UIStore = require("@ui-schema/ui-schema/UIStore");

var _UIMeta = require("@ui-schema/ui-schema/UIMeta");

var _UIRootRenderer = require("@ui-schema/ui-schema/UIRootRenderer");

var _excluded = ["children"];

/**
 * Main Component to create a schema based UI generator
 */
var UIGenerator = function UIGenerator(_ref) {
  var children = _ref.children,
      props = (0, _objectWithoutProperties2["default"])(_ref, _excluded);
  return /*#__PURE__*/_react["default"].createElement(UIProvider, props, /*#__PURE__*/_react["default"].createElement(_UIRootRenderer.UIRootRenderer, {
    schema: props.schema
  }), children);
};

exports.UIGenerator = UIGenerator;

var UIProvider = function UIProvider(_ref2) {
  var children = _ref2.children,
      store = _ref2.store,
      onChange = _ref2.onChange,
      widgets = _ref2.widgets,
      t = _ref2.t,
      showValidity = _ref2.showValidity;
  return /*#__PURE__*/_react["default"].createElement(_UIMeta.UIMetaProvider, {
    widgets: widgets,
    t: t
  }, /*#__PURE__*/_react["default"].createElement(_UIStore.UIStoreProvider, {
    store: store,
    onChange: onChange,
    showValidity: showValidity
  }, children));
};

exports.UIProvider = UIProvider;
