// @ui-schema/ui-schema/Plugins/ReferencingHandler/ReferencingHandler.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ReferencingHandler = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _Translate = require("@ui-schema/ui-schema/Translate");

var _SchemaRootProvider = require("@ui-schema/ui-schema/SchemaRootProvider");

var _ReferencingHandler = require("@ui-schema/ui-schema/Plugins/ReferencingHandler");

var _PluginStack = require("@ui-schema/ui-schema/PluginStack");

var _getSchema = require("@ui-schema/ui-schema/Utils/getSchema");

var _excluded = ["rootContext"],
    _excluded2 = ["schema", "definitions"];

var ReferencingHandler = function ReferencingHandler(_ref) {
  var rootContext = _ref.rootContext,
      props = (0, _objectWithoutProperties2["default"])(_ref, _excluded);
  var baseSchema = props.schema,
      isVirtual = props.isVirtual;

  var _useSchemaRoot = (0, _SchemaRootProvider.useSchemaRoot)(),
      maybeRootSchema = _useSchemaRoot.schema,
      maybeDefinitions = _useSchemaRoot.definitions,
      nestedRootProps = (0, _objectWithoutProperties2["default"])(_useSchemaRoot, _excluded2);

  var isRoot = Boolean((0, _SchemaRootProvider.isRootSchema)(baseSchema) || rootContext || baseSchema.get('definitions') || baseSchema.get('$defs'));
  var definitions = isRoot ? baseSchema.get('definitions') || baseSchema.get('$defs') : maybeDefinitions;

  var _useSchemaRef = (0, _ReferencingHandler.useSchemaRef)(maybeRootSchema, definitions, isRoot, baseSchema),
      schema = _useSchemaRef.schema,
      refsPending = _useSchemaRef.refsPending;

  return refsPending && refsPending.size > 0 ? isVirtual ? null : /*#__PURE__*/_react["default"].createElement(_Translate.Trans, {
    text: 'labels.loading',
    fallback: 'Loading'
  }) : isRoot ? /*#__PURE__*/_react["default"].createElement(_SchemaRootProvider.SchemaRootProvider, (0, _extends2["default"])({}, nestedRootProps, rootContext || {}, {
    id: (0, _getSchema.getSchemaId)(schema),
    schema: schema,
    definitions: definitions
  }), /*#__PURE__*/_react["default"].createElement(_PluginStack.NextPluginRendererMemo, (0, _extends2["default"])({}, props, {
    schema: schema
  }))) : /*#__PURE__*/_react["default"].createElement(_PluginStack.NextPluginRendererMemo, (0, _extends2["default"])({}, props, {
    schema: schema
  }));
};

exports.ReferencingHandler = ReferencingHandler;
