// @ui-schema/ui-schema/Utils/useDebounceValue/useDebounceValue.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useDebounceValue = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireDefault(require("react"));

var useDebounceValue = function useDebounceValue(value, debounceTime, setter) {
  var timer = _react["default"].useRef(undefined);

  var _React$useState = _react["default"].useState({
    changed: false,
    value: undefined
  }),
      _React$useState2 = (0, _slicedToArray2["default"])(_React$useState, 2),
      bounceVal = _React$useState2[0],
      setBounceVal = _React$useState2[1];

  _react["default"].useEffect(function () {
    window.clearTimeout(timer.current);
    setBounceVal({
      changed: false,
      value: value
    });
  }, [value, timer]);

  _react["default"].useEffect(function () {
    if (typeof bounceVal.value === 'undefined' || !bounceVal.changed) return;
    timer.current = window.setTimeout(function () {
      setter(bounceVal.value);
    }, debounceTime);
    return function () {
      return window.clearTimeout(timer.current);
    };
  }, [bounceVal, setBounceVal, debounceTime, timer, setter]);

  var bubbleBounce = _react["default"].useCallback(function (currentVal) {
    if (bounceVal.value === currentVal) return;
    window.clearTimeout(timer.current);
    setter(bounceVal.value);
  }, [setter, bounceVal]);

  var changeBounceVal = _react["default"].useCallback(function (nextVal) {
    setBounceVal({
      changed: true,
      value: nextVal
    });
  }, [setBounceVal]);

  return {
    bubbleBounce: bubbleBounce,
    bounceVal: bounceVal,
    setBounceVal: setBounceVal,
    changeBounceVal: changeBounceVal
  };
};

exports.useDebounceValue = useDebounceValue;
