// @ui-schema/ds-material/esm/Widgets/GenericList/GenericList.js
import _extends from "@babel/runtime/helpers/extends";
import React from 'react';
import { memo } from '@ui-schema/ui-schema/Utils/memo';
import { useUIStore } from '@ui-schema/ui-schema/UIStore';
import { GenericListContent, GenericListFooter, GenericListItem, GenericListItemMore, GenericListItemPos } from '@ui-schema/ds-material/BaseComponents/GenericList';
export const GenericListContentMemo = memo(GenericListContent);
export const GenericList = props => {
  const {
    store
  } = useUIStore();
  const {
    value
  } = (store === null || store === void 0 ? void 0 : store.extractValues(props.storeKeys)) || {}; // extracting and calculating the list size here, not passing down the actual list for performance reasons
  // https://github.com/ui-schema/ui-schema/issues/133

  return /*#__PURE__*/React.createElement(GenericListContentMemo, _extends({}, props, {
    listSize: (value === null || value === void 0 ? void 0 : value.size) || 0,
    ComponentItemPos: GenericListItemPos,
    ComponentItemMore: GenericListItemMore,
    ComponentItem: GenericListItem,
    ComponentFooter: GenericListFooter // e.g. use `Button` instead of `IconButton` for `add-item`
    //btnAddShowLabel

  }));
};
