// @ui-schema/ds-material/widgetsBinding.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.widgets = void 0;

var _react = _interopRequireDefault(require("react"));

var _TextField = require("@ui-schema/ds-material/Widgets/TextField");

var _Select = require("@ui-schema/ds-material/Widgets/Select");

var _OptionsBoolean = require("@ui-schema/ds-material/Widgets/OptionsBoolean");

var _OptionsCheck = require("@ui-schema/ds-material/Widgets/OptionsCheck");

var _OptionsRadio = require("@ui-schema/ds-material/Widgets/OptionsRadio");

var _TextFieldIcon = require("@ui-schema/ds-material/Widgets/TextFieldIcon");

var _SimpleList = require("@ui-schema/ds-material/Widgets/SimpleList");

var _GenericList = require("@ui-schema/ds-material/Widgets/GenericList");

var _NumberSlider = require("@ui-schema/ds-material/Widgets/NumberSlider");

var _Accordions = require("@ui-schema/ds-material/Widgets/Accordions");

var _Grid = require("./Grid");

var _pluginStack = require("./pluginStack");

var _WidgetRenderer = require("@ui-schema/ui-schema/WidgetRenderer");

var _validators = require("@ui-schema/ui-schema/Validators/validators");

var _Widgets = require("@ui-schema/ds-material/Widgets");

var _immutable = require("immutable");

var MyFallbackComponent = function MyFallbackComponent(_ref) {
  var type = _ref.type,
      widget = _ref.widget;
  return /*#__PURE__*/_react["default"].createElement("div", null, /*#__PURE__*/_react["default"].createElement("p", null, /*#__PURE__*/_react["default"].createElement("strong", null, "System Error in Widget!")), /*#__PURE__*/_react["default"].createElement("p", null, /*#__PURE__*/_react["default"].createElement("strong", null, "Type:"), " ", _immutable.List.isList(type) ? type.join(', ') : type || '-'), /*#__PURE__*/_react["default"].createElement("p", null, /*#__PURE__*/_react["default"].createElement("strong", null, "Widget:"), " ", widget || '-'));
};

var widgets = {
  ErrorFallback: MyFallbackComponent,
  RootRenderer: _Grid.RootRenderer,
  GroupRenderer: _Grid.GroupRenderer,
  WidgetRenderer: _WidgetRenderer.WidgetRenderer,
  pluginStack: _pluginStack.pluginStack,
  pluginSimpleStack: _validators.validators,
  types: {
    string: _TextField.StringRenderer,
    "boolean": _OptionsBoolean.BoolRenderer,
    number: _TextField.NumberRenderer,
    integer: _TextField.NumberRenderer
  },
  custom: {
    Accordions: _Accordions.AccordionsRenderer,
    Text: _TextField.TextRenderer,
    StringIcon: _TextFieldIcon.StringIconRenderer,
    TextIcon: _TextFieldIcon.TextIconRenderer,
    NumberIcon: _TextFieldIcon.NumberIconRenderer,
    NumberSlider: _NumberSlider.NumberSlider,
    SimpleList: _SimpleList.SimpleList,
    GenericList: _GenericList.GenericList,
    OptionsCheck: _OptionsCheck.OptionsCheck,
    OptionsRadio: _OptionsRadio.OptionsRadio,
    Select: _Select.Select,
    SelectMulti: _Select.SelectMulti,
    Card: _Widgets.CardRenderer,
    LabelBox: _Widgets.LabelBox,
    FormGroup: _Widgets.FormGroup
  }
};
exports.widgets = widgets;
