// @ui-schema/ds-material/Widgets/Card/Card.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CardRenderer = void 0;

var _react = _interopRequireDefault(require("react"));

var _uiSchema = require("@ui-schema/ui-schema");

var _ObjectRenderer = require("@ui-schema/ui-schema/ObjectRenderer");

var _Card = _interopRequireDefault(require("@material-ui/core/Card"));

var _CardContent = _interopRequireDefault(require("@material-ui/core/CardContent"));

var _Typography = _interopRequireDefault(require("@material-ui/core/Typography"));

var CardRenderer = function CardRenderer(props) {
  var schema = props.schema,
      storeKeys = props.storeKeys,
      ownKey = props.ownKey;
  return /*#__PURE__*/_react["default"].createElement(_Card["default"], {
    elevation: schema.getIn(['view', 'ev']),
    variant: schema.getIn(['view', 'variant']),
    style: schema.getIn(['view', 'bg']) === false ? {
      background: 'transparent'
    } : undefined
  }, /*#__PURE__*/_react["default"].createElement(_CardContent["default"], null, schema.getIn(['view', 'hideTitle']) ? null : /*#__PURE__*/_react["default"].createElement(_Typography["default"], {
    variant: schema.getIn(['view', 'titleVariant']) || 'h5',
    component: schema.getIn(['view', 'titleComp']) || 'p',
    gutterBottom: true
  }, /*#__PURE__*/_react["default"].createElement(_uiSchema.TransTitle, {
    schema: schema,
    storeKeys: storeKeys,
    ownKey: ownKey
  })), /*#__PURE__*/_react["default"].createElement(_ObjectRenderer.ObjectRenderer, props)));
};

exports.CardRenderer = CardRenderer;
