// @ui-schema/ds-material/Component/TitleBoxRead/TitleBoxRead.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TitleBoxRead = void 0;

var _Typography = _interopRequireDefault(require("@material-ui/core/Typography"));

var _uiSchema = require("@ui-schema/ui-schema");

var _Box = _interopRequireDefault(require("@material-ui/core/Box"));

var _react = _interopRequireDefault(require("react"));

var TitleBoxRead = function TitleBoxRead(_ref) {
  var schema = _ref.schema,
      storeKeys = _ref.storeKeys,
      errors = _ref.errors,
      valid = _ref.valid,
      hideTitle = _ref.hideTitle,
      hasInfo = _ref.hasInfo,
      InfoRenderer = _ref.InfoRenderer;
  return /*#__PURE__*/_react["default"].createElement(_Box["default"], {
    style: {
      display: 'flex',
      opacity: 0.75
    }
  }, hideTitle ? null : /*#__PURE__*/_react["default"].createElement(_Typography["default"], {
    variant: 'caption',
    color: !valid ? 'error' : 'textSecondary'
  }, /*#__PURE__*/_react["default"].createElement(_uiSchema.TransTitle, {
    schema: schema,
    storeKeys: storeKeys,
    ownKey: storeKeys.last()
  })), hasInfo ? /*#__PURE__*/_react["default"].createElement(_Box["default"], null, InfoRenderer && schema !== null && schema !== void 0 && schema.get('info') ? /*#__PURE__*/_react["default"].createElement(InfoRenderer, {
    schema: schema,
    variant: 'icon',
    openAs: 'modal',
    storeKeys: storeKeys,
    valid: valid,
    errors: errors
  }) : null) : null);
};

exports.TitleBoxRead = TitleBoxRead;
