// @ui-schema/ds-material/esm/Widgets/Card/Card.js
import React from 'react';
import { TransTitle } from '@ui-schema/ui-schema';
import { ObjectRenderer } from '@ui-schema/ui-schema/ObjectRenderer';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
export const CardRenderer = props => {
  const {
    schema,
    storeKeys,
    ownKey
  } = props;
  return /*#__PURE__*/React.createElement(Card, {
    elevation: schema.getIn(['view', 'ev']),
    variant: schema.getIn(['view', 'variant']),
    style: schema.getIn(['view', 'bg']) === false ? {
      background: 'transparent'
    } : undefined
  }, /*#__PURE__*/React.createElement(CardContent, null, schema.getIn(['view', 'hideTitle']) ? null : /*#__PURE__*/React.createElement(Typography, {
    variant: schema.getIn(['view', 'titleVariant']) || 'h5',
    component: schema.getIn(['view', 'titleComp']) || 'p',
    gutterBottom: true
  }, /*#__PURE__*/React.createElement(TransTitle, {
    schema: schema,
    storeKeys: storeKeys,
    ownKey: ownKey
  })), /*#__PURE__*/React.createElement(ObjectRenderer, props)));
};
