// @ui-schema/ds-material/Widgets/LabelBox/LabelBox.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LabelBox = void 0;

var _react = _interopRequireDefault(require("react"));

var _uiSchema = require("@ui-schema/ui-schema");

var _ObjectRenderer = require("@ui-schema/ui-schema/ObjectRenderer");

var _Box = _interopRequireDefault(require("@material-ui/core/Box"));

var _Typography = _interopRequireDefault(require("@material-ui/core/Typography"));

var LabelBox = function LabelBox(props) {
  var schema = props.schema,
      storeKeys = props.storeKeys,
      ownKey = props.ownKey;
  return /*#__PURE__*/_react["default"].createElement(_Box["default"], {
    mt: schema.getIn(['view', 'mt']),
    mr: schema.getIn(['view', 'mr']),
    mb: schema.getIn(['view', 'mb']),
    ml: schema.getIn(['view', 'ml']),
    pt: schema.getIn(['view', 'pt']),
    pr: schema.getIn(['view', 'pr']),
    pb: schema.getIn(['view', 'pb']),
    pl: schema.getIn(['view', 'pl'])
  }, schema.getIn(['view', 'hideTitle']) ? null : /*#__PURE__*/_react["default"].createElement(_Typography["default"], {
    gutterBottom: true,
    variant: schema.getIn(['view', 'titleVariant']) || 'h5',
    component: schema.getIn(['view', 'titleComp']) || 'p'
  }, /*#__PURE__*/_react["default"].createElement(_uiSchema.TransTitle, {
    schema: schema,
    storeKeys: storeKeys,
    ownKey: ownKey
  })), /*#__PURE__*/_react["default"].createElement(_ObjectRenderer.ObjectRenderer, props));
};

exports.LabelBox = LabelBox;
