// @ui-schema/ds-material/Component/LocaleHelperText/LocaleHelperText.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ValidityHelperText = exports.LocaleHelperText = void 0;

var _react = _interopRequireDefault(require("react"));

var _FormHelperText = _interopRequireDefault(require("@material-ui/core/FormHelperText"));

var _Trans = require("@ui-schema/ui-schema/Translate/Trans");

var LocaleHelperText = function LocaleHelperText(_ref) {
  var text = _ref.text,
      schema = _ref.schema,
      context = _ref.context,
      _ref$error = _ref.error,
      error = _ref$error === void 0 ? false : _ref$error;
  return /*#__PURE__*/_react["default"].createElement(_FormHelperText["default"], {
    error: error
  }, /*#__PURE__*/_react["default"].createElement(_Trans.Trans, {
    text: text,
    context: context ? context.set('type', schema.get('type')).set('widget', schema.get('widget')) : undefined
  }));
};

exports.LocaleHelperText = LocaleHelperText;

var ValidityHelperText = function ValidityHelperText(_ref2) {
  var showValidity = _ref2.showValidity,
      errors = _ref2.errors,
      schema = _ref2.schema,
      browserError = _ref2.browserError;
  return schema.get('t') === 'browser' && browserError ? /*#__PURE__*/_react["default"].createElement(_FormHelperText["default"], {
    error: true
  }, browserError) : showValidity && errors && errors.hasError() ? errors.getErrors().keySeq().map(function (type) {
    return errors.getError(type).map(function (err, i) {
      return /*#__PURE__*/_react["default"].createElement(LocaleHelperText, {
        key: type + '.' + i,
        schema: schema,
        error: true,
        text: 'error.' + type,
        context: err
      });
    });
  }).valueSeq() : null;
};

exports.ValidityHelperText = ValidityHelperText;
