// @react-aria/accordion/dist/main.js
var $hhnGa$reactariautils = require("@react-aria/utils");
var $hhnGa$reactariabutton = require("@react-aria/button");
var $hhnGa$reactariaselection = require("@react-aria/selection");

function $parcel$exportWildcard(dest, source) {
  Object.keys(source).forEach(function(key) {
    if (key === 'default' || key === '__esModule' || dest.hasOwnProperty(key)) {
      return;
    }

    Object.defineProperty(dest, key, {
      enumerable: true,
      get: function get() {
        return source[key];
      }
    });
  });

  return dest;
}
function $parcel$export(e, n, v, s) {
  Object.defineProperty(e, n, {get: v, set: s, enumerable: true, configurable: true});
}
var $b655721e1e8aa6ea$exports = {};

$parcel$export($b655721e1e8aa6ea$exports, "useAccordionItem", () => $b655721e1e8aa6ea$export$dfe72e7a237bd184);
$parcel$export($b655721e1e8aa6ea$exports, "useAccordion", () => $b655721e1e8aa6ea$export$35e88b08325f7140);



function $b655721e1e8aa6ea$export$dfe72e7a237bd184(props, state, ref) {
    let { item: item  } = props;
    let buttonId = $hhnGa$reactariautils.useId();
    let regionId = $hhnGa$reactariautils.useId();
    let isDisabled = state.disabledKeys.has(item.key);
    let { itemProps: itemProps  } = $hhnGa$reactariaselection.useSelectableItem({
        selectionManager: state.selectionManager,
        key: item.key,
        ref: ref
    });
    let { buttonProps: buttonProps  } = $hhnGa$reactariabutton.useButton($hhnGa$reactariautils.mergeProps(itemProps, {
        id: buttonId,
        elementType: 'button',
        isDisabled: isDisabled,
        onPress: ()=>state.toggleKey(item.key)
    }), ref);
    let isExpanded = state.expandedKeys.has(item.key);
    return {
        buttonProps: {
            ...buttonProps,
            'aria-expanded': isExpanded,
            'aria-controls': isExpanded ? regionId : undefined
        },
        regionProps: {
            id: regionId,
            role: 'region',
            'aria-labelledby': buttonId
        }
    };
}
function $b655721e1e8aa6ea$export$35e88b08325f7140(props, state, ref) {
    let { listProps: listProps  } = $hhnGa$reactariaselection.useSelectableList({
        ...props,
        ...state,
        allowsTabNavigation: true,
        ref: ref
    });
    return {
        accordionProps: {
            ...listProps,
            tabIndex: undefined
        }
    };
}


$parcel$exportWildcard(module.exports, $b655721e1e8aa6ea$exports);
