// @react-aria/accordion/dist/module.js
import {useId as $1u3Uc$useId, mergeProps as $1u3Uc$mergeProps} from "@react-aria/utils";
import {useButton as $1u3Uc$useButton} from "@react-aria/button";
import {useSelectableItem as $1u3Uc$useSelectableItem, useSelectableList as $1u3Uc$useSelectableList} from "@react-aria/selection";

function $parcel$export(e, n, v, s) {
  Object.defineProperty(e, n, {get: v, set: s, enumerable: true, configurable: true});
}
var $fdb8889d8fd9a904$exports = {};

$parcel$export($fdb8889d8fd9a904$exports, "useAccordionItem", () => $fdb8889d8fd9a904$export$dfe72e7a237bd184);
$parcel$export($fdb8889d8fd9a904$exports, "useAccordion", () => $fdb8889d8fd9a904$export$35e88b08325f7140);



function $fdb8889d8fd9a904$export$dfe72e7a237bd184(props, state, ref) {
    let { item: item  } = props;
    let buttonId = $1u3Uc$useId();
    let regionId = $1u3Uc$useId();
    let isDisabled = state.disabledKeys.has(item.key);
    let { itemProps: itemProps  } = $1u3Uc$useSelectableItem({
        selectionManager: state.selectionManager,
        key: item.key,
        ref: ref
    });
    let { buttonProps: buttonProps  } = $1u3Uc$useButton($1u3Uc$mergeProps(itemProps, {
        id: buttonId,
        elementType: 'button',
        isDisabled: isDisabled,
        onPress: ()=>state.toggleKey(item.key)
    }), ref);
    let isExpanded = state.expandedKeys.has(item.key);
    return {
        buttonProps: {
            ...buttonProps,
            'aria-expanded': isExpanded,
            'aria-controls': isExpanded ? regionId : undefined
        },
        regionProps: {
            id: regionId,
            role: 'region',
            'aria-labelledby': buttonId
        }
    };
}
function $fdb8889d8fd9a904$export$35e88b08325f7140(props, state, ref) {
    let { listProps: listProps  } = $1u3Uc$useSelectableList({
        ...props,
        ...state,
        allowsTabNavigation: true,
        ref: ref
    });
    return {
        accordionProps: {
            ...listProps,
            tabIndex: undefined
        }
    };
}




export {$fdb8889d8fd9a904$export$dfe72e7a237bd184 as useAccordionItem, $fdb8889d8fd9a904$export$35e88b08325f7140 as useAccordion};
