// @ui-schema/ds-material/esm/BaseComponents/GenericList/GenericListItemMore.js
import React from 'react';
import IconButton from '@material-ui/core/IconButton';
import { Trans } from '@ui-schema/ui-schema';
import { AccessTooltipIcon } from '@ui-schema/ds-material/Component/Tooltip';
import Delete from '@material-ui/icons/Delete';
import { Map } from 'immutable';
export const GenericListItemMore = ({
  index,
  listRequired,
  schema,
  onChange,
  storeKeys,
  notDeletable,
  btnSize = 'small'
}) => {
  const readOnly = schema.get('readOnly');
  return /*#__PURE__*/React.createElement(React.Fragment, null, !readOnly && !notDeletable ? /*#__PURE__*/React.createElement(IconButton, {
    onClick: () => onChange({
      storeKeys,
      scopes: ['value', 'internal'],
      type: 'list-item-delete',
      index: index,
      schema,
      required: listRequired
    }),
    size: btnSize,
    style: {
      margin: '0 0 auto 0'
    }
  }, /*#__PURE__*/React.createElement(AccessTooltipIcon, {
    title: /*#__PURE__*/React.createElement(Trans, {
      text: 'labels.remove-item',
      context: Map({
        actionLabels: schema.get('listActionLabels')
      })
    })
  }, /*#__PURE__*/React.createElement(Delete, {
    fontSize: 'inherit',
    style: {
      margin: btnSize === 'small' ? 2 : undefined
    }
  }))) : null);
};
