// @ui-schema/ui-schema/WidgetRenderer/WidgetRenderer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WidgetRenderer = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _Utils = require("@ui-schema/ui-schema/Utils");

var _widgetMatcher = require("@ui-schema/ui-schema/widgetMatcher");

var _immutable = require("immutable");

var _excluded = ["value", "internalValue", "WidgetOverride", "NoWidget", "errors", "onErrors", "requiredList", "currentPluginIndex"];

var WidgetRenderer = function WidgetRenderer(_ref) {
  var value = _ref.value,
      internalValue = _ref.internalValue,
      WidgetOverride = _ref.WidgetOverride,
      NoWidget = _ref.NoWidget,
      errors = _ref.errors,
      onErrors = _ref.onErrors,
      requiredList = _ref.requiredList,
      currentPluginIndex = _ref.currentPluginIndex,
      props = (0, _objectWithoutProperties2["default"])(_ref, _excluded);
  var schema = props.schema,
      widgets = props.widgets,
      isVirtual = props.isVirtual;
  var currentErrors = (0, _Utils.useImmutable)(errors);

  _react["default"].useEffect(function () {
    return onErrors && onErrors(currentErrors);
  }, [onErrors, currentErrors]);

  var schemaType = schema.get('type');
  var widgetName = schema.get('widget');
  var Widget = (0, _widgetMatcher.widgetMatcher)({
    isVirtual: Boolean(isVirtual),
    WidgetOverride: WidgetOverride,
    NoWidget: NoWidget,
    widgetName: widgetName,
    schemaType: schemaType,
    widgets: widgets
  });
  var noExtractValue = !isVirtual && (schemaType === 'array' || schemaType === 'object' || _immutable.List.isList(schemaType) && (schemaType.indexOf('array') !== -1 || schemaType.indexOf('object') !== -1));
  return Widget ? /*#__PURE__*/_react["default"].createElement(Widget, (0, _extends2["default"])({}, props, {
    value: noExtractValue ? undefined : value,
    internalValue: noExtractValue ? undefined : internalValue,
    errors: currentErrors
  })) : null;
};

exports.WidgetRenderer = WidgetRenderer;
