// @ui-schema/ds-material/Widgets/Stepper/Stepper.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Stepper = exports.Step = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _makeStyles = _interopRequireDefault(require("@material-ui/core/styles/makeStyles"));

var _Stepper = _interopRequireDefault(require("@material-ui/core/Stepper"));

var _Step = _interopRequireDefault(require("@material-ui/core/Step"));

var _StepLabel = _interopRequireDefault(require("@material-ui/core/StepLabel"));

var _Typography = _interopRequireDefault(require("@material-ui/core/Typography"));

var _Button = _interopRequireDefault(require("@material-ui/core/Button"));

var _Utils = require("@ui-schema/ui-schema/Utils");

var _UIStore = require("@ui-schema/ui-schema/UIStore");

var _ValidityReporter = require("@ui-schema/ui-schema/ValidityReporter");

var _PluginStack = require("@ui-schema/ui-schema/PluginStack");

var _TransTitle = require("@ui-schema/ui-schema/Translate/TransTitle");

var _excluded = ["schema", "storeKeys", "level"];
var useStyles = (0, _makeStyles["default"])(function (theme) {
  return {
    root: {
      width: '100%'
    },
    button: {
      marginRight: theme.spacing(1)
    },
    instructions: {
      marginTop: theme.spacing(1),
      marginBottom: theme.spacing(1)
    }
  };
});

var Step = function Step(_ref) {
  var schema = _ref.schema,
      storeKeys = _ref.storeKeys,
      level = _ref.level,
      p = (0, _objectWithoutProperties2["default"])(_ref, _excluded);
  return /*#__PURE__*/_react["default"].createElement(_PluginStack.PluginStack, (0, _extends2["default"])({
    schema: schema["delete"]('widget'),
    storeKeys: storeKeys,
    level: level + 1
  }, p));
};

exports.Step = Step;
var Stepper = (0, _UIStore.extractValidity)((0, _Utils.memo)(function (_ref2) {
  var schema = _ref2.schema,
      storeKeys = _ref2.storeKeys,
      validity = _ref2.validity,
      level = _ref2.level;
  if (!schema) return null;

  var _React$useState = _react["default"].useState(false),
      _React$useState2 = (0, _slicedToArray2["default"])(_React$useState, 2),
      showValidity = _React$useState2[0],
      setShowValidity = _React$useState2[1];

  var classes = useStyles();

  var _React$useState3 = _react["default"].useState(0),
      _React$useState4 = (0, _slicedToArray2["default"])(_React$useState3, 2),
      activeStep = _React$useState4[0],
      setActiveStep = _React$useState4[1];

  var steps = schema.get('properties');
  var stepOrder = steps.keySeq();
  var activeStepInvalid = (0, _ValidityReporter.isInvalid)(validity, [stepOrder.get(activeStep)]);

  var handleNext = function handleNext() {
    if (activeStepInvalid) {
      setShowValidity(true);
    } else {
      setShowValidity(false);
      setActiveStep(function (prevActiveStep) {
        return prevActiveStep + 1;
      });
    }
  };

  var handleBack = function handleBack() {
    setActiveStep(function (prevActiveStep) {
      return prevActiveStep - 1;
    });
  };

  var handleReset = function handleReset() {
    setActiveStep(0);
  };

  return /*#__PURE__*/_react["default"].createElement("div", {
    className: classes.root
  }, /*#__PURE__*/_react["default"].createElement(_Stepper["default"], {
    activeStep: activeStep
  }, steps.map(function (step, name) {
    var stepProps = {};
    var labelProps = {};
    return /*#__PURE__*/_react["default"].createElement(_Step["default"], (0, _extends2["default"])({
      key: name
    }, stepProps), /*#__PURE__*/_react["default"].createElement(_StepLabel["default"], labelProps, /*#__PURE__*/_react["default"].createElement(_TransTitle.TransTitle, {
      schema: schema,
      storeKeys: storeKeys,
      ownKey: name
    })));
  }).valueSeq()), /*#__PURE__*/_react["default"].createElement("div", null, activeStep === stepOrder.size ? /*#__PURE__*/_react["default"].createElement("div", null, /*#__PURE__*/_react["default"].createElement(_Typography["default"], {
    className: classes.instructions
  }, "All steps completed - you're finished"), /*#__PURE__*/_react["default"].createElement(_Button["default"], {
    onClick: handleReset,
    className: classes.button
  }, "Reset")) : /*#__PURE__*/_react["default"].createElement("div", null, /*#__PURE__*/_react["default"].createElement(_Typography["default"], {
    className: classes.instructions
  }, /*#__PURE__*/_react["default"].createElement(_TransTitle.TransTitle, {
    schema: schema,
    storeKeys: storeKeys,
    ownKey: stepOrder.get(activeStep)
  })), /*#__PURE__*/_react["default"].createElement(_PluginStack.PluginStack, {
    showValidity: showValidity,
    schema: steps.get(stepOrder.get(activeStep)),
    parentSchema: schema,
    storeKeys: storeKeys.push(stepOrder.get(activeStep)),
    level: level + 1
  }), /*#__PURE__*/_react["default"].createElement("div", {
    style: {
      margin: '24px 0 0 0'
    }
  }, /*#__PURE__*/_react["default"].createElement(_Button["default"], {
    disabled: activeStep === 0,
    onClick: handleBack,
    className: classes.button
  }, "Back"), /*#__PURE__*/_react["default"].createElement(_Button["default"], {
    variant: "contained",
    color: "primary",
    onClick: handleNext,
    className: classes.button
  }, activeStep === steps.size - 1 ? 'Finish' : 'Next')))));
}));
exports.Stepper = Stepper;
