// @ui-schema/ds-material/esm/BaseComponents/Table/TableRowActionDelete.js
import React from 'react';
import IconButton from '@material-ui/core/IconButton';
import Delete from '@material-ui/icons/Delete';
import { Trans } from '@ui-schema/ui-schema';
import { AccessTooltipIcon } from '@ui-schema/ds-material/Component/Tooltip/Tooltip';
import { Map } from 'immutable';
export const TableRowActionDelete = ({
  onChange,
  storeKeys,
  showRows = 0,
  setPage,
  index,
  deleteOnEmpty,
  schema
}) => {
  return /*#__PURE__*/React.createElement(IconButton, {
    color: "inherit",
    onClick: () => {
      onChange({
        type: 'list-item-delete',
        storeKeys: storeKeys.splice(-1, 1),
        scopes: ['value', 'internal'],
        index: index,
        effect: ({
          value
        }) => {
          if (showRows !== -1) {
            setPage(p => {
              const nextPage = (Math.ceil(value.size / showRows) || 1) - 1;
              return p < nextPage ? p : nextPage;
            });
          }
        },
        schema: Map({
          type: 'array'
        }),
        required: deleteOnEmpty
      });
    },
    size: 'small'
  }, /*#__PURE__*/React.createElement(AccessTooltipIcon, {
    title: /*#__PURE__*/React.createElement(Trans, {
      text: 'labels.remove-row',
      context: Map({
        actionLabels: schema === null || schema === void 0 ? void 0 : schema.get('tableActionLabels')
      })
    })
  }, /*#__PURE__*/React.createElement(Delete, {
    fontSize: 'inherit',
    style: {
      margin: 2
    }
  })));
};
