// @ui-schema/ui-schema/esm/UIGenerator/UIGenerator.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
const _excluded = ["children"];
import React from 'react';
import { UIStoreProvider } from '@ui-schema/ui-schema/UIStore';
import { UIMetaProvider } from '@ui-schema/ui-schema/UIMeta';
import { UIRootRenderer } from '@ui-schema/ui-schema/UIRootRenderer';

/**
 * Main Component to create a schema based UI generator
 */
export const UIGenerator = _ref => {
  let {
    children
  } = _ref,
      props = _objectWithoutProperties(_ref, _excluded);

  return /*#__PURE__*/React.createElement(UIProvider, props, /*#__PURE__*/React.createElement(UIRootRenderer, {
    schema: props.schema
  }), children);
};
export const UIProvider = ({
  children,
  store,
  onChange,
  widgets,
  t,
  showValidity
}) => {
  return /*#__PURE__*/React.createElement(UIMetaProvider, {
    widgets: widgets,
    t: t
  }, /*#__PURE__*/React.createElement(UIStoreProvider, {
    store: store,
    onChange: onChange,
    showValidity: showValidity
  }, children));
};
