// @ui-schema/ds-material/WidgetsRead/WidgetTextFieldRead/WidgetTextFieldRead.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TextRendererRead = exports.StringRendererRead = exports.NumberRendererRead = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _Box = _interopRequireDefault(require("@material-ui/core/Box"));

var _Typography = _interopRequireDefault(require("@material-ui/core/Typography"));

var _LocaleHelperText = require("@ui-schema/ds-material/Component/LocaleHelperText/LocaleHelperText");

var _uiSchema = require("@ui-schema/ui-schema");

var _TitleBoxRead = require("@ui-schema/ds-material/Component/TitleBoxRead");

var _excluded = ["schema"];

var StringRendererRead = function StringRendererRead(_ref) {
  var multiline = _ref.multiline,
      storeKeys = _ref.storeKeys,
      schema = _ref.schema,
      value = _ref.value,
      showValidity = _ref.showValidity,
      valid = _ref.valid,
      errors = _ref.errors,
      style = _ref.style,
      onClick = _ref.onClick,
      widgets = _ref.widgets;
  var hideTitle = Boolean(schema.getIn(['view', 'hideTitle']));
  var InfoRenderer = widgets === null || widgets === void 0 ? void 0 : widgets.InfoRenderer;
  var lines = multiline && typeof value === 'string' ? value.split('\n') : [];
  var hasInfo = Boolean(InfoRenderer && (schema === null || schema === void 0 ? void 0 : schema.get('info')));

  var _useUIMeta = (0, _uiSchema.useUIMeta)(),
      readDense = _useUIMeta.readDense;

  return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement(_Box["default"], {
    onClick: onClick,
    style: style
  }, /*#__PURE__*/_react["default"].createElement(_TitleBoxRead.TitleBoxRead, {
    hideTitle: hideTitle,
    hasInfo: hasInfo,
    InfoRenderer: InfoRenderer,
    valid: valid,
    errors: errors,
    storeKeys: storeKeys,
    schema: schema
  }), multiline ? typeof value === 'string' ? lines.map(function (line, i) {
    return /*#__PURE__*/_react["default"].createElement(_Typography["default"], {
      key: i,
      gutterBottom: i < lines.length - 1,
      variant: readDense ? 'body2' : 'body1'
    }, line);
  }) : /*#__PURE__*/_react["default"].createElement(_Typography["default"], null, /*#__PURE__*/_react["default"].createElement("span", {
    style: {
      opacity: 0.65
    }
  }, "-")) : /*#__PURE__*/_react["default"].createElement(_Typography["default"], {
    variant: readDense ? 'body2' : 'body1'
  }, typeof value === 'string' || typeof value === 'number' ? value : /*#__PURE__*/_react["default"].createElement("span", {
    style: {
      opacity: 0.65
    }
  }, "-")), /*#__PURE__*/_react["default"].createElement(_LocaleHelperText.ValidityHelperText, {
    errors: errors,
    showValidity: showValidity,
    schema: schema
  })));
};

exports.StringRendererRead = StringRendererRead;

var TextRendererRead = function TextRendererRead(_ref2) {
  var schema = _ref2.schema,
      props = (0, _objectWithoutProperties2["default"])(_ref2, _excluded);
  return /*#__PURE__*/_react["default"].createElement(StringRendererRead, (0, _extends2["default"])({}, props, {
    schema: schema,
    multiline: true
  }));
};

exports.TextRendererRead = TextRendererRead;

var NumberRendererRead = function NumberRendererRead(props) {
  return /*#__PURE__*/_react["default"].createElement(StringRendererRead, (0, _extends2["default"])({}, props, {
    type: 'number'
  }));
};

exports.NumberRendererRead = NumberRendererRead;
