// @ui-schema/ds-material/esm/Widgets/LabelBox/LabelBox.js
import React from 'react';
import { TransTitle } from '@ui-schema/ui-schema';
import { ObjectRenderer } from '@ui-schema/ui-schema/ObjectRenderer';
import Box from '@material-ui/core/Box';
import Typography from '@material-ui/core/Typography';
export const LabelBox = props => {
  const {
    schema,
    storeKeys,
    ownKey
  } = props;
  return /*#__PURE__*/React.createElement(Box, {
    mt: schema.getIn(['view', 'mt']),
    mr: schema.getIn(['view', 'mr']),
    mb: schema.getIn(['view', 'mb']),
    ml: schema.getIn(['view', 'ml']),
    pt: schema.getIn(['view', 'pt']),
    pr: schema.getIn(['view', 'pr']),
    pb: schema.getIn(['view', 'pb']),
    pl: schema.getIn(['view', 'pl'])
  }, schema.getIn(['view', 'hideTitle']) ? null : /*#__PURE__*/React.createElement(Typography, {
    gutterBottom: true,
    variant: schema.getIn(['view', 'titleVariant']) || 'h5',
    component: schema.getIn(['view', 'titleComp']) || 'p'
  }, /*#__PURE__*/React.createElement(TransTitle, {
    schema: schema,
    storeKeys: storeKeys,
    ownKey: ownKey
  })), /*#__PURE__*/React.createElement(ObjectRenderer, props));
};
