// @ui-schema/ds-material/Component/ListButton/ListButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListButton = void 0;

var _dsMaterial = require("@ui-schema/ds-material");

var _Button = _interopRequireDefault(require("@material-ui/core/Button"));

var _IconButton = _interopRequireDefault(require("@material-ui/core/IconButton"));

var _react = _interopRequireDefault(require("react"));

var ListButton = function ListButton(_ref) {
  var onClick = _ref.onClick,
      Icon = _ref.Icon,
      title = _ref.title,
      btnSize = _ref.btnSize,
      btnVariant = _ref.btnVariant,
      btnColor = _ref.btnColor,
      style = _ref.style,
      showLabel = _ref.showLabel;
  return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, !showLabel && Icon ? /*#__PURE__*/_react["default"].createElement(_IconButton["default"], {
    onClick: onClick,
    size: btnSize,
    color: btnColor,
    style: style
  }, /*#__PURE__*/_react["default"].createElement(_dsMaterial.AccessTooltipIcon, {
    title: title
  }, /*#__PURE__*/_react["default"].createElement(Icon, {
    fontSize: 'inherit',
    style: {
      margin: btnSize === 'small' ? 2 : undefined
    }
  }))) : /*#__PURE__*/_react["default"].createElement(_Button["default"], {
    onClick: onClick,
    size: btnSize,
    color: btnColor,
    variant: btnVariant,
    style: style,
    startIcon: Icon ? /*#__PURE__*/_react["default"].createElement(Icon, {
      fontSize: 'inherit'
    }) : undefined
  }, title));
};

exports.ListButton = ListButton;
