// @ui-schema/ds-material/WidgetsRead/WidgetOneOfRead/WidgetOneOfRead.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WidgetOneOfRead = void 0;

var _react = _interopRequireDefault(require("react"));

var _immutable = require("immutable");

var _uiSchema = require("@ui-schema/ui-schema");

var _LocaleHelperText = require("@ui-schema/ds-material/Component/LocaleHelperText/LocaleHelperText");

var _Box = _interopRequireDefault(require("@material-ui/core/Box"));

var _TitleBoxRead = require("@ui-schema/ds-material/Component/TitleBoxRead");

var _Typography = _interopRequireDefault(require("@material-ui/core/Typography"));

var checkActive = function checkActive(list, name) {
  return list && list.contains && typeof list.contains(name) !== 'undefined' ? list.contains(name) : false;
};

var OneOfArrayValuesBase = function OneOfArrayValuesBase(_ref) {
  var oneOfValues = _ref.oneOfValues,
      value = _ref.value,
      dense = _ref.dense;
  var vals = oneOfValues === null || oneOfValues === void 0 ? void 0 : oneOfValues.filter(function (v) {
    return checkActive(value, v.get('const'));
  });
  return vals !== null && vals !== void 0 && vals.size ? vals.map(function (oneOfSchema, i) {
    return /*#__PURE__*/_react["default"].createElement(_Typography["default"], {
      variant: dense ? 'body2' : 'body1',
      style: {
        paddingRight: i < vals.size - 1 ? 4 : 0
      },
      key: oneOfSchema.get('const')
    }, /*#__PURE__*/_react["default"].createElement(_uiSchema.Trans, {
      schema: oneOfSchema.get('t'),
      text: oneOfSchema.get('title') || oneOfSchema.get('const'),
      context: (0, _immutable.Map)({
        'relative': (0, _immutable.List)(['title'])
      }),
      fallback: oneOfSchema.get('title') || (0, _uiSchema.beautifyKey)(oneOfSchema.get('const'), oneOfSchema.get('tt'))
    }), i < vals.size - 1 ? ', ' : '');
  }).valueSeq() : /*#__PURE__*/_react["default"].createElement(_Typography["default"], {
    variant: dense ? 'body2' : 'body1',
    style: {
      opacity: 0.65
    }
  }, "-");
};

var OneOfArrayValues = (0, _uiSchema.extractValue)((0, _uiSchema.memo)(OneOfArrayValuesBase));

var WidgetOneOfRead = function WidgetOneOfRead(_ref2) {
  var schema = _ref2.schema,
      storeKeys = _ref2.storeKeys,
      showValidity = _ref2.showValidity,
      valid = _ref2.valid,
      errors = _ref2.errors,
      widgets = _ref2.widgets;
  var hideTitle = schema.getIn(['view', 'hideTitle']);
  var oneOfVal = schema.getIn(['items', 'oneOf']);
  var InfoRenderer = widgets === null || widgets === void 0 ? void 0 : widgets.InfoRenderer;
  var hasInfo = Boolean(InfoRenderer && (schema === null || schema === void 0 ? void 0 : schema.get('info')));
  if (!oneOfVal) return null;

  var _useUIMeta = (0, _uiSchema.useUIMeta)(),
      readDense = _useUIMeta.readDense;

  return /*#__PURE__*/_react["default"].createElement(_Box["default"], null, /*#__PURE__*/_react["default"].createElement(_TitleBoxRead.TitleBoxRead, {
    hideTitle: hideTitle,
    hasInfo: hasInfo,
    InfoRenderer: InfoRenderer,
    valid: valid,
    errors: errors,
    storeKeys: storeKeys,
    schema: schema
  }), /*#__PURE__*/_react["default"].createElement(_Box["default"], {
    style: {
      display: 'flex',
      flexWrap: 'wrap'
    }
  }, /*#__PURE__*/_react["default"].createElement(OneOfArrayValues, {
    oneOfValues: oneOfVal,
    storeKeys: storeKeys,
    schema: schema,
    dense: Boolean(readDense)
  })), /*#__PURE__*/_react["default"].createElement(_LocaleHelperText.ValidityHelperText, {
    errors: errors,
    showValidity: showValidity,
    schema: schema
  }));
};

exports.WidgetOneOfRead = WidgetOneOfRead;
