// @ui-schema/ds-material/esm/Component/TitleBoxRead/TitleBoxRead.js
import Typography from '@material-ui/core/Typography';
import { TransTitle } from '@ui-schema/ui-schema';
import Box from '@material-ui/core/Box';
import React from 'react';
export const TitleBoxRead = ({
  schema,
  storeKeys,
  errors,
  valid,
  hideTitle,
  hasInfo,
  InfoRenderer
}) => {
  return /*#__PURE__*/React.createElement(Box, {
    style: {
      display: 'flex',
      opacity: 0.75
    }
  }, hideTitle ? null : /*#__PURE__*/React.createElement(Typography, {
    variant: 'caption',
    color: !valid ? 'error' : 'textSecondary'
  }, /*#__PURE__*/React.createElement(TransTitle, {
    schema: schema,
    storeKeys: storeKeys,
    ownKey: storeKeys.last()
  })), hasInfo ? /*#__PURE__*/React.createElement(Box, null, InfoRenderer && schema !== null && schema !== void 0 && schema.get('info') ? /*#__PURE__*/React.createElement(InfoRenderer, {
    schema: schema,
    variant: 'icon',
    openAs: 'modal',
    storeKeys: storeKeys,
    valid: valid,
    errors: errors
  }) : null) : null);
};
