// @ui-schema/ds-material/esm/Component/LocaleHelperText/LocaleHelperText.js
import React from 'react';
import FormHelperText from '@material-ui/core/FormHelperText';
import { Trans } from '@ui-schema/ui-schema/Translate/Trans';
export const LocaleHelperText = ({
  text,
  schema,
  context,
  error = false
}) => {
  return /*#__PURE__*/React.createElement(FormHelperText, {
    error: error
  }, /*#__PURE__*/React.createElement(Trans, {
    text: text,
    context: context ? context.set('type', schema.get('type')).set('widget', schema.get('widget')) : undefined
  }));
};
export const ValidityHelperText = ({
  showValidity,
  errors,
  schema,
  browserError
}) => schema.get('t') === 'browser' && browserError ? /*#__PURE__*/React.createElement(FormHelperText, {
  error: true
}, browserError) : showValidity && errors && errors.hasError() ? errors.getErrors().keySeq().map(type => errors.getError(type).map((err, i) => /*#__PURE__*/React.createElement(LocaleHelperText, {
  key: type + '.' + i,
  schema: schema,
  error: true,
  text: 'error.' + type,
  context: err
}))).valueSeq() : null;
