// @ui-schema/ds-material/BaseComponents/GenericList/GenericListItem.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GenericListItemBase = exports.GenericListItem = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _Grid = _interopRequireDefault(require("@material-ui/core/Grid"));

var _Divider = _interopRequireDefault(require("@material-ui/core/Divider"));

var _uiSchema = require("@ui-schema/ui-schema");

var _immutable = require("immutable");

var _excluded = ["ComponentPos", "ComponentMore", "spacing", "btnSize"];

var GenericListItemBase = function GenericListItemBase(_ref) {
  var _itemsSchema$get;

  var ComponentPos = _ref.ComponentPos,
      ComponentMore = _ref.ComponentMore,
      _ref$spacing = _ref.spacing,
      spacing = _ref$spacing === void 0 ? 2 : _ref$spacing,
      btnSize = _ref.btnSize,
      props = (0, _objectWithoutProperties2["default"])(_ref, _excluded);
  var index = props.index,
      listSize = props.listSize,
      schema = props.schema,
      storeKeys = props.storeKeys,
      schemaKeys = props.schemaKeys,
      level = props.level,
      showValidity = props.showValidity;
  var ownKeys = storeKeys.push(index);
  var itemsSchema = schema.get('items');
  return [/*#__PURE__*/_react["default"].createElement(_Grid["default"], {
    key: 'a',
    item: true,
    xs: 12,
    style: {
      display: 'flex'
    }
  }, /*#__PURE__*/_react["default"].createElement(_Grid["default"], {
    container: true,
    spacing: spacing,
    wrap: 'nowrap'
  }, ComponentPos ? /*#__PURE__*/_react["default"].createElement(_Grid["default"], {
    item: true,
    style: {
      display: 'flex',
      flexDirection: 'column',
      flexShrink: 0
    }
  }, /*#__PURE__*/_react["default"].createElement(ComponentPos, props)) : null, itemsSchema && _immutable.List.isList(itemsSchema) ? // tuples in root level not possible
  // was wrong implementation <= 0.2.2
  null : (0, _uiSchema.schemaTypeToDistinct)(itemsSchema.get('type')) === 'array' && itemsSchema.get('items') ? /*#__PURE__*/_react["default"].createElement(_Grid["default"], {
    item: true,
    style: {
      display: 'flex',
      flexDirection: 'column',
      flexGrow: 2
    }
  }, /*#__PURE__*/_react["default"].createElement(_Grid["default"], {
    container: true,
    spacing: 2
  }, (_itemsSchema$get = itemsSchema.get('items')) === null || _itemsSchema$get === void 0 ? void 0 : _itemsSchema$get.map(function (item, j) {
    return /*#__PURE__*/_react["default"].createElement(_uiSchema.PluginStack, {
      key: j,
      showValidity: showValidity,
      schemaKeys: schemaKeys === null || schemaKeys === void 0 ? void 0 : schemaKeys.push('items').push('items').push(j),
      storeKeys: ownKeys.push(j),
      schema: item,
      parentSchema: schema,
      level: level + 1
    });
  }).valueSeq())) : /*#__PURE__*/_react["default"].createElement(_uiSchema.PluginStack, {
    showValidity: showValidity,
    schema: itemsSchema,
    parentSchema: schema,
    storeKeys: ownKeys,
    level: level + 1,
    schemaKeys: schemaKeys === null || schemaKeys === void 0 ? void 0 : schemaKeys.push('items')
  }), ComponentMore ? /*#__PURE__*/_react["default"].createElement(_Grid["default"], {
    item: true,
    style: {
      display: 'flex',
      flexShrink: 0
    }
  }, /*#__PURE__*/_react["default"].createElement(ComponentMore, (0, _extends2["default"])({}, props, {
    btnSize: btnSize
  }))) : null)), index < listSize - 1 ? /*#__PURE__*/_react["default"].createElement(_Divider["default"], {
    key: 'b',
    style: {
      width: '100%'
    }
  }) : null];
};

exports.GenericListItemBase = GenericListItemBase;
var GenericListItem = (0, _uiSchema.memo)(GenericListItemBase);
exports.GenericListItem = GenericListItem;
