// @ui-schema/ds-material/BaseComponents/GenericList/GenericListItemMore.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GenericListItemMore = void 0;

var _react = _interopRequireDefault(require("react"));

var _IconButton = _interopRequireDefault(require("@material-ui/core/IconButton"));

var _uiSchema = require("@ui-schema/ui-schema");

var _Tooltip = require("@ui-schema/ds-material/Component/Tooltip");

var _Delete = _interopRequireDefault(require("@material-ui/icons/Delete"));

var _immutable = require("immutable");

var GenericListItemMore = function GenericListItemMore(_ref) {
  var index = _ref.index,
      listRequired = _ref.listRequired,
      schema = _ref.schema,
      onChange = _ref.onChange,
      storeKeys = _ref.storeKeys,
      notDeletable = _ref.notDeletable,
      _ref$btnSize = _ref.btnSize,
      btnSize = _ref$btnSize === void 0 ? 'small' : _ref$btnSize;
  var readOnly = schema.get('readOnly');
  return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, !readOnly && !notDeletable ? /*#__PURE__*/_react["default"].createElement(_IconButton["default"], {
    onClick: function onClick() {
      return onChange({
        storeKeys: storeKeys,
        scopes: ['value', 'internal'],
        type: 'list-item-delete',
        index: index,
        schema: schema,
        required: listRequired
      });
    },
    size: btnSize,
    style: {
      margin: '0 0 auto 0'
    }
  }, /*#__PURE__*/_react["default"].createElement(_Tooltip.AccessTooltipIcon, {
    title: /*#__PURE__*/_react["default"].createElement(_uiSchema.Trans, {
      text: 'labels.remove-item',
      context: (0, _immutable.Map)({
        actionLabels: schema.get('listActionLabels')
      })
    })
  }, /*#__PURE__*/_react["default"].createElement(_Delete["default"], {
    fontSize: 'inherit',
    style: {
      margin: btnSize === 'small' ? 2 : undefined
    }
  }))) : null);
};

exports.GenericListItemMore = GenericListItemMore;
