// @ui-schema/ds-material/BaseComponents/Table/TablePaginationActions.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TablePaginationActions = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _styles = require("@material-ui/core/styles");

var _IconButton = _interopRequireDefault(require("@material-ui/core/IconButton"));

var _FirstPage = _interopRequireDefault(require("@material-ui/icons/FirstPage"));

var _KeyboardArrowLeft = _interopRequireDefault(require("@material-ui/icons/KeyboardArrowLeft"));

var _KeyboardArrowRight = _interopRequireDefault(require("@material-ui/icons/KeyboardArrowRight"));

var _LastPage = _interopRequireDefault(require("@material-ui/icons/LastPage"));

var _dsMaterial = require("@ui-schema/ds-material");

var _uiSchema = require("@ui-schema/ui-schema");

var _excluded = ["noFirstPageButton"],
    _excluded2 = ["noLastPageButton"];
var useStyles1 = (0, _styles.makeStyles)(function (theme) {
  return (0, _styles.createStyles)({
    root: {
      flexShrink: 0,
      marginLeft: theme.spacing(2.5)
    }
  });
});

var TablePaginationActions = function TablePaginationActions(props) {
  var classes = useStyles1();
  var theme = (0, _styles.useTheme)();
  var count = props.count,
      page = props.page,
      rowsPerPage = props.rowsPerPage,
      onPageChange = props.onPageChange,
      backIconButtonPropsTmp = props.backIconButtonProps,
      nextIconButtonPropsTmp = props.nextIconButtonProps;
  var btnSize = backIconButtonPropsTmp === null || backIconButtonPropsTmp === void 0 ? void 0 : backIconButtonPropsTmp.size; // @ts-ignore

  var _ref = backIconButtonPropsTmp || {},
      noFirstPageButton = _ref.noFirstPageButton,
      backIconButtonProps = (0, _objectWithoutProperties2["default"])(_ref, _excluded); // @ts-ignore


  var _ref2 = nextIconButtonPropsTmp || {},
      noLastPageButton = _ref2.noLastPageButton,
      nextIconButtonProps = (0, _objectWithoutProperties2["default"])(_ref2, _excluded2);

  var handleFirstPageButtonClick = function handleFirstPageButtonClick(event) {
    onPageChange(event, 0);
  };

  var handleBackButtonClick = function handleBackButtonClick(event) {
    onPageChange(event, page - 1);
  };

  var handleNextButtonClick = function handleNextButtonClick(event) {
    onPageChange(event, page + 1);
  };

  var handleLastPageButtonClick = function handleLastPageButtonClick(event) {
    onPageChange(event, Math.max(0, Math.ceil(count / rowsPerPage) - 1));
  };

  var iconStyle = {
    padding: btnSize === 'small' ? 2 : undefined
  };
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: classes.root
  }, noFirstPageButton ? null : /*#__PURE__*/_react["default"].createElement(_IconButton["default"], {
    onClick: handleFirstPageButtonClick,
    disabled: page === 0,
    size: btnSize
  }, /*#__PURE__*/_react["default"].createElement(_dsMaterial.AccessTooltipIcon, {
    title: /*#__PURE__*/_react["default"].createElement(_uiSchema.Trans, {
      text: 'pagination.first-page'
    })
  }, theme.direction === 'rtl' ? /*#__PURE__*/_react["default"].createElement(_LastPage["default"], {
    style: iconStyle
  }) : /*#__PURE__*/_react["default"].createElement(_FirstPage["default"], {
    style: iconStyle
  }))), /*#__PURE__*/_react["default"].createElement(_IconButton["default"], (0, _extends2["default"])({
    onClick: handleBackButtonClick,
    disabled: page === 0,
    size: btnSize
  }, backIconButtonProps), /*#__PURE__*/_react["default"].createElement(_dsMaterial.AccessTooltipIcon, {
    title: /*#__PURE__*/_react["default"].createElement(_uiSchema.Trans, {
      text: 'pagination.prev-page'
    })
  }, theme.direction === 'rtl' ? /*#__PURE__*/_react["default"].createElement(_KeyboardArrowRight["default"], {
    style: iconStyle
  }) : /*#__PURE__*/_react["default"].createElement(_KeyboardArrowLeft["default"], {
    style: iconStyle
  }))), /*#__PURE__*/_react["default"].createElement(_IconButton["default"], (0, _extends2["default"])({
    onClick: handleNextButtonClick,
    disabled: page >= Math.ceil(count / rowsPerPage) - 1,
    style: {
      padding: btnSize === 'small' ? 2 : undefined
    },
    size: btnSize
  }, nextIconButtonProps), /*#__PURE__*/_react["default"].createElement(_dsMaterial.AccessTooltipIcon, {
    title: /*#__PURE__*/_react["default"].createElement(_uiSchema.Trans, {
      text: 'pagination.next-page'
    })
  }, theme.direction === 'rtl' ? /*#__PURE__*/_react["default"].createElement(_KeyboardArrowLeft["default"], {
    style: iconStyle
  }) : /*#__PURE__*/_react["default"].createElement(_KeyboardArrowRight["default"], {
    style: iconStyle
  }))), noLastPageButton ? null : /*#__PURE__*/_react["default"].createElement(_IconButton["default"], {
    onClick: handleLastPageButtonClick,
    disabled: page >= Math.ceil(count / rowsPerPage) - 1,
    size: btnSize,
    style: {
      padding: btnSize === 'small' ? 2 : undefined
    }
  }, /*#__PURE__*/_react["default"].createElement(_dsMaterial.AccessTooltipIcon, {
    title: /*#__PURE__*/_react["default"].createElement(_uiSchema.Trans, {
      text: 'pagination.last-page'
    })
  }, theme.direction === 'rtl' ? /*#__PURE__*/_react["default"].createElement(_FirstPage["default"], {
    style: iconStyle
  }) : /*#__PURE__*/_react["default"].createElement(_LastPage["default"], {
    style: iconStyle
  }))));
};

exports.TablePaginationActions = TablePaginationActions;
