// @ui-schema/ds-material/BaseComponents/Table/TableContext.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useTable = exports.TableContext = void 0;
exports.withTable = withTable;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _uiSchema = require("@ui-schema/ui-schema");

// @ts-ignore
var TableContext = /*#__PURE__*/_react["default"].createContext({});

exports.TableContext = TableContext;

function withTable(Component) {
  var WithTable = function WithTable(p) {
    var _useTable = useTable(),
        errors = _useTable.errors,
        valid = _useTable.valid;

    return /*#__PURE__*/_react["default"].createElement(Component, (0, _extends2["default"])({
      valid: valid,
      errors: errors
    }, p));
  };

  WithTable.displayName = "ExtractValidity(".concat((0, _uiSchema.getDisplayName)(Component), ")");
  return WithTable;
}

var useTable = function useTable() {
  return _react["default"].useContext(TableContext);
};

exports.useTable = useTable;
