// @ui-schema/ds-material/BaseComponents/Table/TableRowActionDelete.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TableRowActionDelete = void 0;

var _react = _interopRequireDefault(require("react"));

var _IconButton = _interopRequireDefault(require("@material-ui/core/IconButton"));

var _Delete = _interopRequireDefault(require("@material-ui/icons/Delete"));

var _uiSchema = require("@ui-schema/ui-schema");

var _Tooltip = require("@ui-schema/ds-material/Component/Tooltip/Tooltip");

var _immutable = require("immutable");

var TableRowActionDelete = function TableRowActionDelete(_ref) {
  var onChange = _ref.onChange,
      storeKeys = _ref.storeKeys,
      _ref$showRows = _ref.showRows,
      showRows = _ref$showRows === void 0 ? 0 : _ref$showRows,
      setPage = _ref.setPage,
      index = _ref.index,
      deleteOnEmpty = _ref.deleteOnEmpty,
      schema = _ref.schema;
  return /*#__PURE__*/_react["default"].createElement(_IconButton["default"], {
    color: "inherit",
    onClick: function onClick() {
      onChange({
        type: 'list-item-delete',
        storeKeys: storeKeys.splice(-1, 1),
        scopes: ['value', 'internal'],
        index: index,
        effect: function effect(_ref2) {
          var value = _ref2.value;

          if (showRows !== -1) {
            setPage(function (p) {
              var nextPage = (Math.ceil(value.size / showRows) || 1) - 1;
              return p < nextPage ? p : nextPage;
            });
          }
        },
        schema: (0, _immutable.Map)({
          type: 'array'
        }),
        required: deleteOnEmpty
      });
    },
    size: 'small'
  }, /*#__PURE__*/_react["default"].createElement(_Tooltip.AccessTooltipIcon, {
    title: /*#__PURE__*/_react["default"].createElement(_uiSchema.Trans, {
      text: 'labels.remove-row',
      context: (0, _immutable.Map)({
        actionLabels: schema === null || schema === void 0 ? void 0 : schema.get('tableActionLabels')
      })
    })
  }, /*#__PURE__*/_react["default"].createElement(_Delete["default"], {
    fontSize: 'inherit',
    style: {
      margin: 2
    }
  })));
};

exports.TableRowActionDelete = TableRowActionDelete;
