// @ui-schema/ds-material/Widgets/OptionsBoolean/OptionsBoolean.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BoolRenderer = void 0;

var _react = _interopRequireDefault(require("react"));

var _FormControlLabel = _interopRequireDefault(require("@material-ui/core/FormControlLabel"));

var _Switch = _interopRequireDefault(require("@material-ui/core/Switch"));

var _makeStyles = _interopRequireDefault(require("@material-ui/core/styles/makeStyles"));

var _colors = require("@material-ui/core/colors");

var _uiSchema = require("@ui-schema/ui-schema");

var _Component = require("@ui-schema/ds-material/Component");

var switchStyle = (0, _makeStyles["default"])(function (theme) {
  return {
    switchBase: {
      color: function color(_ref) {
        var error = _ref.error;
        return error ? theme.palette.error.main : theme.palette.type === 'dark' ? _colors.grey[400] : _colors.grey[50];
      }
    },
    checked: {},
    track: {
      backgroundColor: function backgroundColor(_ref2) {
        var error = _ref2.error;
        return error ? theme.palette.error.dark : theme.palette.type === 'dark' ? _colors.grey[500] : _colors.grey[300];
      }
    }
  };
});

var BoolRenderer = function BoolRenderer(_ref3) {
  var ownKey = _ref3.ownKey,
      value = _ref3.value,
      _onChange = _ref3.onChange,
      schema = _ref3.schema,
      storeKeys = _ref3.storeKeys,
      showValidity = _ref3.showValidity,
      valid = _ref3.valid,
      required = _ref3.required,
      errors = _ref3.errors,
      _ref3$labelledBy = _ref3.labelledBy,
      labelledBy = _ref3$labelledBy === void 0 ? undefined : _ref3$labelledBy;
  var currentVal = Boolean(value);
  var classes = switchStyle({
    error: !valid && showValidity
  });

  var control = /*#__PURE__*/_react["default"].createElement(_Switch["default"], {
    classes: classes,
    required: required,
    checked: currentVal,
    disabled: schema.get('readOnly'),
    inputProps: {
      'aria-labelledby': labelledBy
    },
    onChange: function onChange() {
      return _onChange({
        storeKeys: storeKeys,
        scopes: ['value'],
        type: 'update',
        updater: function updater(_ref4) {
          var storeValue = _ref4.value;
          return {
            value: !storeValue
          };
        },
        schema: schema,
        required: required
      });
    }
  });

  return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, schema.getIn(['view', 'hideTitle']) ? control : /*#__PURE__*/_react["default"].createElement(_FormControlLabel["default"], {
    disabled: schema.get('readOnly'),
    control: control,
    label: /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement(_uiSchema.TransTitle, {
      schema: schema,
      storeKeys: storeKeys,
      ownKey: ownKey
    }), required ? ' *' : '')
  }), /*#__PURE__*/_react["default"].createElement(_Component.ValidityHelperText, {
    errors: errors,
    showValidity: showValidity,
    schema: schema
  }));
};

exports.BoolRenderer = BoolRenderer;
