// @ui-schema/ds-material/Widgets/GenericList/GenericList.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GenericListContentMemo = exports.GenericList = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _memo = require("@ui-schema/ui-schema/Utils/memo");

var _UIStore = require("@ui-schema/ui-schema/UIStore");

var _GenericList = require("@ui-schema/ds-material/BaseComponents/GenericList");

var GenericListContentMemo = (0, _memo.memo)(_GenericList.GenericListContent);
exports.GenericListContentMemo = GenericListContentMemo;

var GenericList = function GenericList(props) {
  var _useUIStore = (0, _UIStore.useUIStore)(),
      store = _useUIStore.store;

  var _ref = (store === null || store === void 0 ? void 0 : store.extractValues(props.storeKeys)) || {},
      value = _ref.value; // extracting and calculating the list size here, not passing down the actual list for performance reasons
  // https://github.com/ui-schema/ui-schema/issues/133


  return /*#__PURE__*/_react["default"].createElement(GenericListContentMemo, (0, _extends2["default"])({}, props, {
    listSize: (value === null || value === void 0 ? void 0 : value.size) || 0,
    ComponentItemPos: _GenericList.GenericListItemPos,
    ComponentItemMore: _GenericList.GenericListItemMore,
    ComponentItem: _GenericList.GenericListItem,
    ComponentFooter: _GenericList.GenericListFooter // e.g. use `Button` instead of `IconButton` for `add-item`
    //btnAddShowLabel

  }));
};

exports.GenericList = GenericList;
