// @ui-schema/ui-schema/UIStore/UIStoreProvider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UIConfigProvider = void 0;
exports.UIStoreProvider = UIStoreProvider;
exports.useUI = void 0;
exports.useUIConfig = useUIConfig;
exports.useUIStore = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _UIStoreActions = require("@ui-schema/ui-schema/UIStoreActions");

var _excluded = ["children"],
    _excluded2 = ["children", "showValidity", "onChange", "store"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

// @ts-ignore
var UIStoreContextObj = /*#__PURE__*/_react["default"].createContext({}); // @ts-ignore


var UIConfigContextObj = /*#__PURE__*/_react["default"].createContext({});

var UIConfigProvider = function UIConfigProvider(_ref) {
  var children = _ref.children,
      props = (0, _objectWithoutProperties2["default"])(_ref, _excluded);

  var value = _react["default"].useMemo(function () {
    return _objectSpread({}, props);
  }, (0, _toConsumableArray2["default"])(Object.values(props)));

  return /*#__PURE__*/_react["default"].createElement(UIConfigContextObj.Provider, {
    value: value
  }, children);
};

exports.UIConfigProvider = UIConfigProvider;

function UIStoreProvider(_ref2) {
  var children = _ref2.children,
      showValidity = _ref2.showValidity,
      onChange = _ref2.onChange,
      store = _ref2.store,
      props = (0, _objectWithoutProperties2["default"])(_ref2, _excluded2);

  var ctx = _react["default"].useMemo(function () {
    return {
      showValidity: showValidity,
      store: store
    };
  }, [showValidity, store]);

  return /*#__PURE__*/_react["default"].createElement(UIStoreContextObj.Provider, {
    value: ctx
  }, /*#__PURE__*/_react["default"].createElement(UIConfigProvider, props, /*#__PURE__*/_react["default"].createElement(_UIStoreActions.UIStoreActionsProvider, {
    onChange: onChange
  }, children)));
}
/**
 * @deprecated
 */


var useUI = function useUI() {
  return useUIStore();
}; // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-constraint


exports.useUI = useUI;

var useUIStore = function useUIStore() {
  return _react["default"].useContext(UIStoreContextObj);
};

exports.useUIStore = useUIStore;

function useUIConfig() {
  // @ts-ignore
  return _react["default"].useContext(UIConfigContextObj);
}
