// @ui-schema/ds-material/Widgets/TableAdvanced/TableAdvanced.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TableAdvancedBase = exports.TableAdvanced = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _uiSchema = require("@ui-schema/ui-schema");

var _excluded = ["showValidity", "schema", "level"];

var TableAdvancedBase = function TableAdvancedBase(_ref) {
  var showValidity = _ref.showValidity,
      schema = _ref.schema,
      level = _ref.level,
      props = (0, _objectWithoutProperties2["default"])(_ref, _excluded);
  var storeKeys = props.storeKeys;
  var readOnly = schema.get('readOnly');
  return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement(_uiSchema.PluginStack, {
    showValidity: showValidity,
    storeKeys: storeKeys.push('data'),
    schema: schema.getIn(['properties', 'data']),
    parentSchema: schema,
    level: level + 1,
    readOnly: readOnly,
    noGrid: true
  }));
};

exports.TableAdvancedBase = TableAdvancedBase;
var TableAdvanced = (0, _uiSchema.extractValue)((0, _uiSchema.memo)(TableAdvancedBase));
exports.TableAdvanced = TableAdvanced;
