// @ui-schema/ds-material/WidgetsRead/WidgetChipsRead/WidgetChipsRead.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WidgetChipsReadBase = exports.WidgetChipsRead = void 0;

var _react = _interopRequireDefault(require("react"));

var _uiSchema = require("@ui-schema/ui-schema");

var _LocaleHelperText = require("@ui-schema/ds-material/Component/LocaleHelperText/LocaleHelperText");

var _Chip = _interopRequireDefault(require("@material-ui/core/Chip"));

var _Box = _interopRequireDefault(require("@material-ui/core/Box"));

var _immutable = require("immutable");

var _TitleBoxRead = require("@ui-schema/ds-material/Component/TitleBoxRead");

var _Typography = _interopRequireDefault(require("@material-ui/core/Typography"));

var WidgetChipsReadBase = function WidgetChipsReadBase(_ref) {
  var storeKeys = _ref.storeKeys,
      schema = _ref.schema,
      value = _ref.value,
      showValidity = _ref.showValidity,
      errors = _ref.errors,
      valid = _ref.valid,
      widgets = _ref.widgets;

  var _useUIMeta = (0, _uiSchema.useUIMeta)(),
      readDense = _useUIMeta.readDense;

  if (!schema) return null;
  var oneOfVal = schema.getIn(['items', 'oneOf']);
  if (!oneOfVal) return null;
  var hideTitle = schema.getIn(['view', 'hideTitle']);
  var InfoRenderer = widgets === null || widgets === void 0 ? void 0 : widgets.InfoRenderer;
  var hasInfo = Boolean(InfoRenderer && (schema === null || schema === void 0 ? void 0 : schema.get('info')));
  var currentValue = typeof value !== 'undefined' ? value : (0, _immutable.List)(schema.get('default')) || (0, _immutable.List)();
  var oneOfValues = oneOfVal.filter(function (v) {
    return (currentValue === null || currentValue === void 0 ? void 0 : currentValue.indexOf(v.get('const'))) !== -1;
  });
  return /*#__PURE__*/_react["default"].createElement(_Box["default"], null, /*#__PURE__*/_react["default"].createElement(_TitleBoxRead.TitleBoxRead, {
    hideTitle: hideTitle,
    hasInfo: hasInfo,
    InfoRenderer: InfoRenderer,
    valid: valid,
    errors: errors,
    storeKeys: storeKeys,
    schema: schema
  }), /*#__PURE__*/_react["default"].createElement(_Box["default"], {
    style: {
      display: 'flex',
      flexWrap: 'wrap'
    }
  }, oneOfValues !== null && oneOfValues !== void 0 && oneOfValues.size ? oneOfValues.map(function (oneOfSchema) {
    return /*#__PURE__*/_react["default"].createElement(_Chip["default"], {
      key: oneOfSchema.get('const'),
      label: /*#__PURE__*/_react["default"].createElement(_uiSchema.Trans, {
        schema: oneOfSchema.get('t'),
        text: oneOfSchema.get('title') || oneOfSchema.get('const'),
        context: (0, _immutable.Map)({
          'relative': (0, _immutable.List)(['title'])
        }),
        fallback: oneOfSchema.get('title') || (0, _uiSchema.beautifyKey)(oneOfSchema.get('const'), oneOfSchema.get('tt'))
      }),
      style: {
        marginRight: 4,
        marginBottom: readDense ? 2 : 4
      },
      size: !readDense && schema.getIn(['view', 'size']) === 'medium' ? 'medium' : 'small',
      variant: 'default',
      color: 'primary'
    });
  }).valueSeq() : /*#__PURE__*/_react["default"].createElement(_Typography["default"], {
    variant: readDense ? 'body2' : 'body1',
    style: {
      opacity: 0.65
    }
  }, "-")), /*#__PURE__*/_react["default"].createElement(_LocaleHelperText.ValidityHelperText, {
    errors: errors,
    showValidity: showValidity,
    schema: schema
  }));
};

exports.WidgetChipsReadBase = WidgetChipsReadBase;
var WidgetChipsRead = (0, _uiSchema.extractValue)((0, _uiSchema.memo)(WidgetChipsReadBase));
exports.WidgetChipsRead = WidgetChipsRead;
