// @ui-schema/ui-schema/Validators/EmailValidator/EmailValidator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.emailValidator = exports.ERROR_EMAIL_INVALID = void 0;

var _schemaTypeIs = require("@ui-schema/ui-schema/Utils/schemaTypeIs");

var ERROR_EMAIL_INVALID = 'email-invalid';
exports.ERROR_EMAIL_INVALID = ERROR_EMAIL_INVALID;
var emailValidator = {
  // @ts-ignore
  should: function should(_ref) {
    var schema = _ref.schema,
        value = _ref.value;
    return Boolean(schema && typeof value === 'string' && (0, _schemaTypeIs.schemaTypeIs)(schema.get('type'), 'string') && schema.get('format') === 'email');
  },
  // @ts-ignore
  handle: function handle(_ref2) {
    var value = _ref2.value,
        errors = _ref2.errors,
        valid = _ref2.valid;

    /*
     *  regex from https://emailregex.com/
     */
    if (!/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/.test(value)) {
      var _errors;

      valid = false;
      errors = (_errors = errors) === null || _errors === void 0 ? void 0 : _errors.addError(ERROR_EMAIL_INVALID);
    }

    return {
      errors: errors,
      valid: valid
    };
  }
};
exports.emailValidator = emailValidator;
