// @ui-schema/ui-schema/Validators/ValueValidator/valueValidatorConst.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.valueValidatorConst = exports.validateConst = exports.ERROR_CONST_MISMATCH = void 0;

var _immutable = require("immutable");

var ERROR_CONST_MISMATCH = 'const-mismatch';
exports.ERROR_CONST_MISMATCH = ERROR_CONST_MISMATCH;

var validateConst = function validateConst(_const, value) {
  return typeof _const === 'undefined' || typeof value === 'undefined' ? // todo: add deep check with List/Map.equals
  true : value === _const;
};

exports.validateConst = validateConst;
var valueValidatorConst = {
  should: function should(_ref) {
    var schema = _ref.schema,
        value = _ref.value;

    var _const = schema.get('const');

    return typeof _const !== 'undefined' && typeof value !== 'undefined';
  },
  handle: function handle(_ref2) {
    var schema = _ref2.schema,
        value = _ref2.value,
        errors = _ref2.errors,
        valid = _ref2.valid;

    if (!validateConst(schema.get('const'), value)) {
      valid = false;
      errors = errors.addError(ERROR_CONST_MISMATCH, (0, _immutable.Map)({
        "const": schema.get('const')
      }));
    }

    return {
      errors: errors,
      valid: valid
    };
  }
};
exports.valueValidatorConst = valueValidatorConst;
