// @ui-schema/ds-material/Widgets/Table/Table.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Table = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _immutable = require("immutable");

var _uiSchema = require("@ui-schema/ui-schema");

var _TableRenderer = require("@ui-schema/ds-material/BaseComponents/Table/TableRenderer");

var _TableFooter = require("@ui-schema/ds-material/BaseComponents/Table/TableFooter");

var _TableHeader = require("@ui-schema/ds-material/BaseComponents/Table/TableHeader");

var _TableRowRenderer = require("@ui-schema/ds-material/BaseComponents/Table/TableRowRenderer");

var TableHeaderMemo = (0, _uiSchema.memo)(_TableHeader.TableHeader);
var TableRowRendererMemo = (0, _uiSchema.memo)(_TableRowRenderer.TableRowRenderer);
var rowsPerPageDefault = (0, _immutable.List)([5, 10, 25]);

var Table = function Table(props) {
  return /*#__PURE__*/_react["default"].createElement(_TableRenderer.TableRenderer, (0, _extends2["default"])({}, props, {
    TableRowRenderer: TableRowRendererMemo,
    TableFooter: _TableFooter.TableFooter,
    TableHeader: TableHeaderMemo,
    rowsPerPage: props.schema.getIn(['view', 'rowsPerPage']) || rowsPerPageDefault,
    rowsShowAll: props.schema.getIn(['view', 'rowsShowAll'])
  }));
};

exports.Table = Table;
